dmqtt/signals.py

```diff
@@ -8,16 +8,32 @@
 
 logger = logging.getLogger(__name__)
 
-connect = django.dispatch.Signal(providing_args=["userdata", "flags", "rc"])
-message = django.dispatch.Signal(providing_args=["userdata", "msg"])
-
+try:
+    connect = django.dispatch.Signal(providing_args=["userdata", "flags", "rc"])
+    message = django.dispatch.Signal(providing_args=["userdata", "msg"])
+except TypeError:
+    # for Django < 3.0
+    connect = django.dispatch.Signal()
+    message = django.dispatch.Signal()
+
+def convert_wildcards(mqtt_topic):
+    """Convert MQTT wildcards to regex pattern.
+    '#' becomes '.*' (multi-level)
+    '+' becomes '[^/]+' (single-level)
+    """
+    pattern = re.escape(mqtt_topic)
+    pattern = pattern.replace('\\#', '.*')
+    pattern = pattern.replace('\\+', '[^/]+')
+    return '^' + pattern + '$'
 
 def topic(matcher, as_json=True, **extras):
     def wrap(func):
         @functools.wraps(func)
         def inner(msg, **kwargs):
-            if fnmatch.fnmatch(msg.topic, matcher):
-                logger.debug("Matched %s for %s", matcher, func)
+            nonlocal matcher
+            regex_pattern = convert_wildcards(matcher)
+            if re.match(regex_pattern, msg.topic):
+                logger.debug("Matched %s for %s", regex_pattern, func)
```

Comment:
```suggestion
            if regex_pattern.match(msg.topic):
                logger.debug("Matched %s for %s", regex_pattern, func)
```

Now this line uses our compiled matcher